module/builder/collection/builder.go

```diff
@@ -195,13 +195,29 @@ func (b *Builder) BuildOn(parentID flow.Identifier, setter func(*flow.Header) er
 		minRefID := refChainFinalizedID
 
 		var transactions []*flow.TransactionBody
+		var totalByteSize uint
+		var totalGas uint64
 		for _, tx := range b.transactions.All() {
 
 			// if we have reached maximum number of transactions, stop
 			if uint(len(transactions)) >= b.config.MaxCollectionSize {
 				break
 			}
 
+			// we can break cause
+			// max byte size per tx << the max collection byte size
+			// to make it more effective in the future we can continue adding smaller ones
+			if totalByteSize+tx.ByteSize() > b.config.MaxCollectionByteSize {
```

Comment:
Nitpicking - `.ByteSize()` requires iterating over collection so its not optimal to call it here, and then again at the bottom of the function. 
We can add and check if sum is over a threshold. Not to mention we save one addition per iteration!

Reply:
I changed it to call it only once but I think saving that one addition might not worth the simplicity of the code reading right now.